File: pypp_cli/test_dir/cpp/src/default_dict/operations.cpp
#include "src/default_dict/operations.h"
#include "py_dict.h"
#include "py_dict_default.h"
#include "py_str.h"
#include "pypp_util/print.h"

namespace me {
void default_dict_operations_fn() {
    pypp::print(pypp::PyStr("DEFAULT DICT OPERATIONS RESULTS:"));
    auto a = pypp::PyDefaultDict<int, pypp::PyStr>::str_factory();
    a.update({{1, pypp::PyStr("one")},
              {2, pypp::PyStr("two")},
              {3, pypp::PyStr("three")}});
    if (a.contains(1)) {
        pypp::print(pypp::PyStr("1 in a"));
    }
    if (!a.contains(4)) {
        pypp::print(pypp::PyStr("4 not in a"));
    }
    pypp::print(a.len());
    pypp::print(a.min());
    pypp::print(a.max());
    pypp::PyDict<int, pypp::PyStr> b = {{1, pypp::PyStr("one")},
                                        {2, pypp::PyStr("two")},
                                        {3, pypp::PyStr("three")}};
    if (a == b) {
        pypp::print(pypp::PyStr("is equal"));
    }
    pypp::PyDict<int, pypp::PyStr> c = {{1, pypp::PyStr("one")},
                                        {2, pypp::PyStr("two")},
                                        {3, pypp::PyStr("thee")}};
    if (a != c) {
        pypp::print(pypp::PyStr("is not equal"));
    }
}

} // namespace me

File: pypp_cli/test_dir/src/default_dict/operations.py
from collections import defaultdict
from pypp_python import auto


def default_dict_operations_fn():
    print("DEFAULT DICT OPERATIONS RESULTS:")
    a: auto = defaultdict[int, str](str)
    a.update({1: "one", 2: "two", 3: "three"})
    if 1 in a:
        print("1 in a")
    if 4 not in a:
        print("4 not in a")
    print(len(a))
    print(min(a))
    print(max(a))
    b: dict[int, str] = {1: "one", 2: "two", 3: "three"}
    # note: when comparing defaultdict with dict, the default dict must be on the left
    # side of the comparison or a compilation error occurs.
    if a == b:
        print("is equal")
    c: dict[int, str] = {1: "one", 2: "two", 3: "thee"}
    if a != c:
        print("is not equal")
